Fabric.Terminology.SqlServer/Persistence/SqlClientTermValueSetRepository.cs

```diff
@@ -9,135 +9,163 @@
             this.logger = logger;
+            this.uowManager = uowManager;
+            this.cacheManager = cacheManager;
         }
 
-        protected ClientTermContext ClientTermContext => this.clientTermContext.Value;
-
         public Maybe<IValueSet> GetValueSet(Guid valueSetGuid)
         {
-            var desc = this.ClientTermContext.ValueSetDescriptions.AsNoTracking()
-                .SingleOrDefault(x => x.ValueSetGUID == valueSetGuid);
+            return this.uowManager.GetValueSetDescriptionDto(valueSetGuid)
+                .Select(
+                    dto =>
+                        {
+                            var factory = new ValueSetBackingItemFactory();
+                            var item = factory.Build(dto);
+                            ((ValueSetBackingItem)item).IsCustom = true;
+                            var codes = this.GetCodes(item.ValueSetGuid);
+                            var counts = this.GetCodeCounts(item.ValueSetGuid);
+
+                            return new ValueSet(item, codes, counts) as IValueSet;
+                        });
+        }
 
-            if (desc == null)
+        public Attempt<IValueSet> Add(IValueSet valueSet)
+        {
+            if (!EnsureIsNew(valueSet))
             {
-                return Maybe.Not;
+                var invalid = new InvalidOperationException("Cannot save an existing value set as a new value set.");
+                return Attempt<IValueSet>.Failed(invalid);
             }
 
-            var factory = new ValueSetBackingItemFactory();
-            var item = factory.Build(desc);
+            var valueSetGuid = valueSet.SetIdsForCustomInsert();
+
+            var descCountOps =
+                CreateOperation(new ValueSetDescriptionBaseDto(valueSet), OperationType.Create)
+                .AppendOperationBatch(
+                    valueSet.CodeCounts.Select(count => new ValueSetCodeCountDto(count)),
+                        OperationType.Create);
+
+            // Insert ValueSetDescriptionBASE and ValueSetCodeCountBASE
+            var uowDescCount = this.uowManager.CreateUnitOfWork(descCountOps);
+            uowDescCount.Commit();
 
-            var codes = this.GetCodes(item.ValueSetGuid);
-            var counts = this.GetCodeCounts(item.ValueSetGuid);
+            // Bulk Insert ValueSetCodeBASE
+            var codesDtos = valueSet.ValueSetCodes.Select(code => new ValueSetCodeDto(code)).ToList();
+            var uowCounts = this.uowManager.CreateBulkCopyUnitOfWork(codesDtos);
+            uowCounts.Commit();
 
-            return new Maybe<IValueSet>(new ValueSet(item, codes, counts));
+            // Get the updated ValueSet
+            var added = this.GetValueSet(valueSetGuid);
+
+            return added.Select(Attempt<IValueSet>.Successful)
+                .Else(
+                    () => Attempt<IValueSet>.Failed(
+                        new ValueSetNotFoundException("Could not retrieved newly saved ValueSet")));
```

Comment:
When possible, include relevant run-time data like the `valueSetGuid` on Exception messages. Details like that can save us a lot of debugging time in the long run.